pkg/agent/controller/agent.go

```diff
@@ -0,0 +1,207 @@
+package controller
+
+import (
+	"fmt"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+)
+
+const (
+	submarinerIpamGlobalIp = "submariner.io/globalIp"
+)
+
+var excludeSvcList = map[string]bool{"kubernetes": true, "openshift": true}
+
+func New(spec *AgentSpecification, config InformerConfigStruct) *Controller {
+	exclusionNSMap := map[string]bool{"kube-system": true, "submariner": true, "openshift": true,
+		"submariner-operator": true, "kubefed-operator": true}
+	for _, v := range spec.ExcludeNS {
+		exclusionNSMap[v] = true
+	}
+	agentController := &Controller{
+		kubeClientSet:    config.KubeClientSet,
+		serviceWorkqueue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "Services"),
+		servicesSynced:   config.ServiceInformer.Informer().HasSynced,
+
+		excludeNamespaces: exclusionNSMap,
+	}
+
+	klog.Info("Setting up event handlers")
+	config.ServiceInformer.Informer().AddEventHandlerWithResyncPeriod(cache.ResourceEventHandlerFuncs{
+		AddFunc: agentController.enqueueService,
+		UpdateFunc: func(old, newObj interface{}) {
+			agentController.handleUpdateService(old, newObj)
+		},
+		DeleteFunc: agentController.handleRemovedService,
+	}, 0)
```

Comment:
I'm not sure what affect passing 0 has but we should just call `AddEventHandler`.

Reply:
0 disables resync. without argument it uses the default specified in informer.

Reply:
right - sorry what I meant was I'm not sure what affect passing 0 has wrt overriding the non-zero setting on the informer. I think perhaps we should configure the resync period in one place at the informer level.  The informer is actually only used here anyway. There's no point  in setting it to 2 min in **main** only to disable it here (if that is indeed the override behavior).